Dispatch exceptions to the sync context (#28936)
dotnet/AspNetCore.Docs: aspnetcore/blazor/components/index.md

> [!IMPORTANT]
> If a Razor component defines an event that's triggered from a background thread, the component might be required to capture and restore the execution context (<xref:System.Threading.ExecutionContext>) at the time the handler is registered. For more information, see [Calling `InvokeAsync(StateHasChanged)` causes page to fallback to default culture (dotnet/aspnetcore #28521)](https://github.com/dotnet/aspnetcore/issues/28521).

:::moniker-end

:::moniker range=">= aspnetcore-8.0"

### Dispatch exceptions to Blazor's synchronization context

> Use `ComponentBase.DispatchExceptionAsync` in Razor components to process exceptions thrown outside of Blazor's synchronization context. This permits the app's code to treat the exceptions as through they're lifecycle method exceptions. Thereafter, Blazor's error handling mechanisms, such as [error boundaries](xref:blazor/fundamentals/handle-errors#error-boundaries), can process the exceptions.

Reviewer: It's not about being outside the sync context. It's about being outside the lifecycle methods. People should use
DispatchExceptionAsync when:

 * Some code *outside the call stack of a lifecycle method* causes an exception, which they catch
 * They want it to be handled as if it was within the lifecycle of a specific component instance (for example, triggering the error boundaries for that component)

This is true independently of whether their code is running on the sync context. The sync context doesn't come into it.

> [!NOTE]
> `ComponentBase.DispatchExceptionAsync` is used in Razor component files (`.razor`) that inherit from <xref:Microsoft.AspNetCore.Components.ComponentBase>. When creating components that [implement <xref:Microsoft.AspNetCore.Components.IComponent> directly](#component-classes), use `RenderHandle.DispatchExceptionAsync`.

> To dispatch exceptions to Blazor's synchronization context, use a `try-catch` block to pass the exception to `DispatchExceptionAsync` and await the result:

Reviewer: This seems a bit odd because:

1. There's not necessarily any need to use `InvokeAsync` here - that depends on what code they are going to call. Whether or not `InvokeAsync` is required is unrelated to the scenario.
2. We probably don't need to explain to people how to catch exceptions - that's a basic C# topic which we can assume readers already know.

Of course you *can* use try/catch in your example, and it makes sense to do so, but it's just odd to point it out in the description as if there's some other way to get exceptions.

```csharp
try
{
    await InvokeAsync(() =>
    {
        ...
    });
}
catch (Exception ex)
{
    await DispatchExceptionAsync(ex);
}
```

For an example use of `DispatchExceptionAsync` API, implement the timer notification example in the [Invoke component methods externally to update state](#invoke-component-methods-externally-to-update-state) section. In a Blazor app, add the following files from the timer notification example and register the services in `Program.cs` as the text explains:

* `TimerService.cs`
* `NotifierService.cs`
* `Pages/ReceiveNotifications.razor`

> The example uses a timer outside of Blazor's synchronization context, where an unhandled exception normally doesn't reach Blazor's synchronization context and isn't processed by Blazor's error handling mechanisms, such as an [error boundary](xref:blazor/fundamentals/handle-errors#error-boundaries).

Reviewer: This is also a bit off-track because it's talking about whether things are on the sync context, which is unrelated to the use for `DispatchExceptionAsync`. You are correct to use something like a timer as an example of something that is not within the call stack of a lifecycle method, though, and that is the relevant thing for `DispatchExceptionAsync`.

> [!NOTE]
> `ComponentBase.DispatchExceptionAsync` is used in Razor component files (`.razor`) that inherit from <xref:Microsoft.AspNetCore.Components.ComponentBase>. When creating components that [implement <xref:Microsoft.AspNetCore.Components.IComponent> directly](#component-classes), use `RenderHandle.DispatchExceptionAsync`.

To dispatch exceptions to Blazor's synchronization context, use a `try-catch` block to pass the exception to `DispatchExceptionAsync` and await the result:

```csharp
try
{
    await InvokeAsync(() =>
    {
        ...
    });
}
catch (Exception ex)
{
    await DispatchExceptionAsync(ex);
}
```

For an example use of `DispatchExceptionAsync` API, implement the timer notification example in the [Invoke component methods externally to update state](#invoke-component-methods-externally-to-update-state) section. The example uses a timer outside of Blazor's synchronization context, where an unhandled exception normally doesn't reach Blazor's synchronization context and isn't processed by Blazor's error handling mechanisms, such as an [error boundary](xref:blazor/fundamentals/handle-errors#error-boundaries).

> First, change the code in `TimerService.cs` to create an artificial exception outside of Blazor's synchronization context. In the `while` loop of `TimerService.cs`, throw an exception when the `elapsedCount` reaches a value of two:

Reviewer 2: I know this is similar to the existing pattern we follow in the docs, where we build on top of an example from earlier in the page, but this model is fragile. In case of a reader following the documentation from the top, they will sure be successful, but if they've landed on this particular section through a link, they will start guessing what `TimerService` is. Hence, I think it'll be good to explicitly call it out like so:

```suggestion
First, change the code in `TimerService.cs` (from an earlier example in this page) to create an artificial exception outside of Blazor's synchronization context. In the `while` loop of `TimerService.cs`, throw an exception when the `elapsedCount` reaches a value of two:
```

Author: Let me try and fix this by making it more explicit in the preceding paragraph. That's why I started by telling the reader there that they would be building on that example. I'll try and resolve this on the next commit.

```csharp
if (elapsedCount == 2)
{
    throw new Exception("I threw an exception! Somebody help me!");
}
```

Place an [error boundary](xref:blazor/fundamentals/handle-errors#error-boundaries) in the app's main layout.

`Shared/MainLayout.razor`:

```razor
<article class="content px-4">
    <ErrorBoundary>
        <ChildContent>
            @Body
        </ChildContent>
        <ErrorContent>
            <p class="alert alert-danger" role="alert">
                Oh, dear! Oh, my! - George Takei
            </p>
        </ErrorContent>
    </ErrorBoundary>
</article>
```

> Normally, the preceding error boundary doesn't react to unhandled exceptions thrown outside of Blazor's synchronization context. If you run the app at this point, the exception is thrown when the elapsed count reaches a value of two, but the error boundary doesn't show the error content.

Reviewer 2: The `Normally` implies, that there is a "trick" to do and it will show, but given the current state of code, it won't. Hence, I think we should remove the `Normally,` part.

```suggestion
If you run the app at this point, the exception is thrown when the elapsed count reaches a value of two, but nothing on the UI changes and the error boundary doesn't show the error content.
```

Author: Sounds good, but we're running on a bit with a looooong sentence. Let's break it it up into separate thoughts (probably better for MT anyway) .....

```suggestion
If you run the app at this point, the exception is thrown when the elapsed count reaches a value of two. However, nothing in the UI changes. The error boundary doesn't show the error content.
```
